nixos:nixos-rebuild: allow switching to a specific generation number

miallo opened the pull request:
###### Motivation for this change
If the user wants to `rollback` to any of the system generations but the last, this is currently not easily doable (see #24374).
In order to unify the interface with `nix-env` a bit more, I propose to add a flag ``nixos-rebuild --generation` to easily go to a specific generation.

I had an issue, where I didn't notice a package breaking ~10 generations ago and while fixing it, I had to go back and forth multiple times. Since I didn't know about
```sh
sudo nix-env --switch-generation 12345 -p /nix/var/nix/profiles/system
sudo /nix/var/nix/profiles/system/bin/switch-to-configuration switch
```
I repeated the rollback `n` times.

This implementation would make it possible to run
```sh
sudo nixos-rebuild switch --generation 12345
```

###### Maintainers
@nbp added `--rollback`
@worldofpeace did the last edits
And I guess @edolstra ?

###### Things done

- [x] Tested manually on local system
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

miallo committed:
nixos-rebuild: Add --generation flag

Run `nixos-rebuild switch --generation 12345` to roll back to generation
`12345`. This makes rolling back to an older generation easier

Also updated the release notes and add a test for the new feature

Co-Authored-By: Jason Yundt <jason@jasonyundt.email>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/394

miallo commented:
I am not sure about having `list-generations` in `nixos-rebuild`, because it isn't strictly related to rebuilding. But on the other hand, having a seperate `nixos-list-generations` command feels a bit overkill as well, looking at the short list of existing `nixos-*` commands... On the other hand one could add parameters for formatting and so on, but I am not sure if that is really needed...

Any opinions @edolstra or others?

miallo commented:
Many ideas for the `list-generations` come from the [script that builds the grub-menu](https://github.com/NixOS/nixpkgs/blob/c14f14eeaf919c914e4dec2ce485a5bdc8dd4fec/nixos/modules/system/boot/loader/generations-dir/generations-dir-builder.sh). Maybe it would be worth, having one script that produces a description for a given generation so that grub, as well as the `list-generations` can utilize it?

edolstra commented:
Maybe this should be its own subcommand? I.e. `nixos-rebuild switch-generation` rather than `switch --switch-generation`, which seems a bit superfluous. 

Or alternatively, `nixos-rebuild [switch|test] --generation`.

miallo commented:
I totally agree with your suggestions and I was a bit unhappy with the doubling of the "switch" as well. The reason I chose it nevertheless is that I took the same flag as `nix-env`...
So far I only used `switch` and I don't know if others would want to `test` old generations. But given the fact that `--rollback` supports both, I guess having a similar interface would be good. I think the `nixos-rebuild [switch|test] --generation` sounds nice.

miallo commented:
Any opinions about the `list-generations` subcommand? I added it to this PR, because with the option to switch to a specific generation I thought it would be nice to have an overview about the main things. I guess a `(current)` tag as in `nix-env --list-generations` might be usefull as well.

edolstra commented:
I think it's okay to add it to `nixos-rebuild` for now. There is an issue somewhere about replacing `nixos-rebuild` with a `nixos` command that would resolve the fact that it's not related to building.

miallo commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -422,11 +426,43 @@ if [ "$action" = dry-build ]; then
+
+    description=""
+    for generation_dir in /nix/var/nix/profiles/system-*-link ; do
+        description="$description\n$(describe_generation "$generation_dir")"
```
I don't particularly like this, but I don't know how to do this in a nicer way...

xaverdh replied:
you could use a bash array, I guess

miallo commented:
Should I clean up the commit history and have one for `list-generation` and one for `--generation`? Or should everything be squashed into one commit?

miallo commented:
Anyone wants to take a look and review this?

xaverdh commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -341,7 +344,7 @@ fi
 # First build Nix, since NixOS may require a newer version than the
 # current one.
-if [ -n "$rollback" -o "$action" = dry-build ]; then
+if [ -n "$rollback" -o -n "$generation" -o "$action" = dry-build ]; then
```
I know this has been here before (for rollback), but why do we use a new nix version to switch to an *old* config?

xaverdh replied:
I mean ideally we would actually use the *old* nix (the one that was used to build the *old* system). We can't really do that due to global state like the nix db I guess, but still using an even newer nix feels wrong.
People might want to rollback, because an update broke important part of their system (internet connection, maybe something with the filesystem tooling etc.) so we should probably do as little as possible to get back to the old state?

miallo replied:
I totally get your point. You could pass `--no-build-nix` with the current implementation to avoid this. I don't know if that should be the default behaviour...

Mic92 replied:
Schema upgrades in nix database or the new (experimental) nix profiles could potentially be a problem.

miallo replied:
Did I get it right: It should be removed for both `--generation` and `--rollback`?

xaverdh replied:
In my opinion yes. I can see no reason to update nix on such an operation.

miallo commented on nixos/modules/installer/tools/nixos-rebuild.sh:
```diff
@@ -440,17 +441,18 @@ if [ "$action" = list-generations ]; then
     done
-    current_generation="$(generation_from_dir $(readlink /nix/var/nix/profiles/system))"
-    echo -e "$description" |
-        sed 's/^\<'"$current_generation"'\>.*/&  (current)/' | # add current generation tag to line
```
I also removed the `sed` in favour of the `describe_generation` echoing the "(current)", because I think it is much more obvious what hapens

miallo commented:
Any further ideas on the `list-generation`?
- Could some information on flakes (if available) be useful? `nixos-version --revision` output? Truncated? (side note: the error message for "no hash found" is written to `stdout` and not `stderr`)
- Is the `configuration-name` of interest? Do people use it in a way that makes sense to show it?
- What about showing it in a pager? The output might be long...

I think it would be good to have a sort of stable output format for the future (or go the `git` way and have `--porcelain` versions of it?)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-to-show-in-list-of-nixos-generations/11216/1

miallo commented on nixos/modules/installer/tools/nixos-version.sh:
```diff
@@ -1,23 +1,36 @@
 #! @runtimeShell@
 
+if [ $# -eq 0 ]; then
```
This is a potentially breaking change, since by default every other parameter would output the default. Nevertheless, it is much better to be explicit and throw errors, if the user mistyped.

miallo commented on nixos/modules/installer/tools/nixos-version.sh:
```diff
@@ -1,23 +1,36 @@
     echo "@revision@"
     ;;
+  --configurationRevision)
+    if ! [[ @configurationRevision@ =~ ^[0-9a-f]+$ ]]; then
```
Not quite the right check, since the configurationRevision could be set to arbitrary text (including uppercase letters). Maybe check if empty instead?

miallo commented:
For an easier review I have [another branch](https://github.com/miallo/nixpkgs/tree/nixos-rebuild/clean-commits), wich is identical in content, but has only three commits. Each commit is separated by different content: `nixos-version`, `list-generations` and `--generation`. This way you don't have to jump mentally between loosely related topics.

A couple of questions I am not sure about for `list-generations`:
- Is there any other useful information to show?
- Is the order fine?
- Truncate configuration revision (maybe if hash only)? Takes a lot of space if set to the flake-commit...
- Put the `(current)` tag in it's own column? This might reduce readability if there is no configuration revision for the current and the surrounding lines...
- Is there a more straight forward way to produce the `--json` output?
- In the man-page I only referenced the `--json` flag in the subsection of `list-generations`, since including it "top-level" looked as if it was for all options and it is already hard to understand all the available flags. Should this become it's own entry nevertheless? It currently isn't documented too well...

xaverdh commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -53,6 +55,14 @@ while [ "$#" -gt 0 ]; do
         rollback=1
         ;;
+      --generation)
+        if [ $# -eq 1 ]; then
```
This should be
```sh
if [ $# -eq 0 ]; then
```
The code above (where `i` is assigned) shifts the index before you enter this code block.

miallo replied:
Good point...

xaverdh commented on nixos/doc/manual/man-nixos-rebuild.xml:
```diff
@@ -425,6 +449,21 @@
+    <listitem>
+     <para>
+      Instead of building a new configuration as specified by
+      <filename>/etc/nixos/configuration.nix</filename>, roll back to a
```
nit: `switch` is probably more appropriate terminology here (instead of `roll back`), since the generation could very well be the currently active one, if it was added with `nixos-rebuild boot`.

miallo replied:
I am all ears for the nitty-gritty details ;)

xaverdh commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -422,11 +434,76 @@ if [ "$action" = dry-build ]; then
+        nixos_version="$(cat "$generation_dir/nixos-version" 2> /dev/null || echo "Unknown")"
+
+        kernel_dir="$(dirname "$(realpath "$generation_dir/kernel")")"
+        kernel_version="$(ls "$kernel_dir/lib/modules")"
```
This might not exist in the unlikely case where the kernel is compiled without modules, so it probably should at least have `|| :` so that it doesn't exit.

miallo replied:
Should it be empty or default to `unknown`? My first instinkt was unknown, but empty might be more appropriate, the more I think about it :thinking:

miallo replied:
Now you got me thinking: the install-grub skript [seems to assume](https://github.com/NixOS/nixpkgs/blob/6e2f594665a0ddd36b1cb764cb42474a0a8f08e5/nixos/modules/system/boot/loader/grub/install-grub.pl#L543-L546) `lib/modules` exists if `nixos-version` isn't there. Is that a valid assumption?

miallo replied:
Something is strange in the grub-script: if `nixos-version` doesn't exist it will [skip this generation](https://github.com/NixOS/nixpkgs/blob/6e2f594665a0ddd36b1cb764cb42474a0a8f08e5/nixos/modules/system/boot/loader/grub/install-grub.pl#L538-L540), so the [later check](https://github.com/NixOS/nixpkgs/blob/57231d7524d94dd992bb8726af8c941d11e2bee3/nixos/modules/system/boot/loader/grub/install-grub.pl#L544) should always succed, right? :thinking:

xaverdh replied:
Hmm it seems both the grub and the systemd-boot builders assume that the modules directory exists. The code that [assembles the modules](https://github.com/xaverdh/nixpkgs/blob/8aaba2790df85abff58bc7815076aef84743b349/pkgs/build-support/kernel/modules-closure.sh#L8) on the other hand takes extra care to handle this case.
I wonder if there is a better way to pass around the kernel version.

xaverdh replied:
and you are right, it should always succeed..

miallo replied:
[here](https://github.com/NixOS/nixpkgs/blob/57231d7524d94dd992bb8726af8c941d11e2bee3/nixos/modules/system/boot/loader/grub/install-grub.pl#L507) is another instance which isn't guarded, but I have read that perl is a bit relaxed, when it comes to `array out of bound` situations, so I guess it is fine

xaverdh replied:
So the actual thing that gets linked in toplevel is `config.system.modulesTree` which is assembled from the `kernel` and `config.boot.extraModulePackages` via the [aggregator](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/kmod/aggregator.nix). This also checks if the modules directory exists, so in theory it *can* be missing I think. Probably not too many people compile kernels without modules.

xaverdh replied:
Ah never mind, I was confused again. Thats the `kernel-modules` symlink. The bootloader generators read the actual `kernel` symlink..

xaverdh replied:
> [here](https://github.com/NixOS/nixpkgs/blob/57231d7524d94dd992bb8726af8c941d11e2bee3/nixos/modules/system/boot/loader/grub/install-grub.pl#L507) is another instance which isn't guarded, but I have read that perl is a bit relaxed, when it comes to `array out of bound` situations, so I guess it is fine?

not sure, I am no expert on perl

xaverdh replied:
I think as far as this pr is concerned, you can just fall back to "Unknown" (or something like that) for now.

miallo commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -422,11 +434,76 @@ if [ "$action" = dry-build ]; then
+            column --separator "," --table --table-columns "Generation,Build-date,NixOS version,Kernel,Configuration Revision" |
+                ${PAGER:less}
+        else
+            tr '\n' ',' | sed 's/,$//' | cat <(echo '{') - <(echo '}')
```
Should the `--json` output be an array instead of an object? That way it might be easier to get the information for relative generations (`echo '[{"a":"1"},{"a":"2"}]' | jq last`), but as an object it is easier for absolute numbers (`echo '{"1": {"a":"3"},"2":{"a":"4"}}' | jq '."1"'`).

xaverdh replied:
I would favor an array, if we make sure, that the generations are ordered correctly (I think currently this is by accident / filesystem ordering, right?), and an object otherwise. But no strong opinion on my part.

miallo replied:
No, the output of `find` is [`sort`ed](https://github.com/NixOS/nixpkgs/blob/ea6cac67219731b5afddcfc86f9a72b86526db22/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L489). In a previous implementation I used globbing without special sorting, but this didn't play nicely if you had profiles like `test` and `test-1` - you probably still remembered this solution...

xaverdh commented:
The `column` binary is from `util-linux`, right? Can you add `util-linux` as dependency [here](https://github.com/NixOS/nixpkgs/blob/ea6cac67219731b5afddcfc86f9a72b86526db22/pkgs/os-specific/linux/nixos-rebuild/default.nix#L21)?

miallo commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -463,7 +463,12 @@ if [ "$action" = list-generations ]; then
+            build_date="$(date --date="@$(stat "$generation_dir" --format=%W)" "+%a %F %T")"
+        else
+            # jq automatically quotes the output => don't try to quote it in output!
+            build_date="$(stat "$generation_dir" --format=%W | jq 'todate')"
```
I just saw that the `flake.lock` file uses the seconds instead of a human and machine readable ISO 8601 format. My initial Idea was to have it sortable and machine, as well as human readable. but since the raw output isn't made for human readability I would also be fine with the seconds.

Mic92 replied:
I think using the dates from the flake would be also miss leading.

dramforever commented:
Hi, not sure if this is the best place to ask, but I personally would love it if `nixos-rebuild` could implement the following

- (Part of this PR, if I got it right) `--generation {generation number}`
- `--system {path to system closure}`: Switch to a certain system closure, instead of a generation. This would allow one to run the bulid process themselves without `nixos-rebuild`, and just letting `nixos-rebuild` handle the switching part. This can be implemented by just replacing `--switch-generation {generation number}` with `--set {system closure}` in the `nix-env` invocation, so I hope this isn't a big problem. It's a bit tricky to do manually now, as you have seen, and it could be useful in automated scripts.

gilescope commented:
Love to see this PR land!

miallo commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -451,6 +455,80 @@ if [ "$action" = dry-build ]; then
+        no_link_gen="${generation_base%-link}"  # remove the "-link"
+        echo ${no_link_gen##*-} # remove everything before the last dash
+    }
+    describe_generation(){
```
In a different programming language I would much rather have this function generic and then have a transform, depending on if the output is `json` or not. Only I don't know how to do this in bash...

miallo commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -374,7 +383,7 @@ trap cleanup EXIT
 
 # First build Nix, since NixOS may require a newer version than the
 # current one.
-if [ -n "$rollback" -o "$action" = dry-build ]; then
```
That should ideally have been in a separate commit...

miallo reviewed (commented):
Sorry for force-pushing, but I reckoned that no-one reviewed this in the past half year and I think now with concise patches for each individual part of the change it will be easier to review...

xaverdh commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -451,11 +464,85 @@ if [ "$action" = dry-build ]; then
+    }
+
+    find $(dirname $profile) -regex "$profile-[0-9]+-link" |
+        sort |
```
I think something like `sort -rg` is more appropriate here?
At least on my system I get this order:
``
100
101
102
103
104
105
106
107
108
109
110
95
96
97
98
99
``

miallo commented:
Thank you very much @xaverdh for your review! @Mic92 do you also want to take a look at it? :)

SuperSandro2000 commented on nixos/modules/installer/tools/nixos-version.sh:
```diff
@@ -13,6 +13,13 @@ case "$1" in
     fi
     echo "@revision@"
     ;;
+  --configurationRevision)
```
```suggestion
  --configuration-revision)
```
flags are usually this style IIRC.

SuperSandro2000 commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -503,11 +516,85 @@ if [ "$action" = dry-build ]; then
+    describe_generation(){
+        generation_dir="$1"
+        generation_number="$(generation_from_dir "$generation_dir")"
+        nixos_version="$(cat "$generation_dir/nixos-version" 2> /dev/null || echo "Unknown")"
```
I think bash can directly read the file without catting.

miallo replied:
Good point! I didn't think of that!
Though I have troubles ignoring the error with the redirect (I tried `a="$(<does_not_exist.txt 2> /dev/null)"` and it still prints "No such file or directory"). Is that even something to consider with the `nixos-version` file? I might have been a bit too defensive in the first place, but I don't know...

SuperSandro2000 commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -503,11 +516,85 @@ if [ "$action" = dry-build ]; then
+        nixos_version="$(cat "$generation_dir/nixos-version" 2> /dev/null || echo "Unknown")"
+
+        kernel_dir="$(dirname "$(realpath "$generation_dir/kernel")")"
+        kernel_version="$(ls "$kernel_dir/lib/modules" || echo Unknown)"
```
```suggestion
        kernel_version="$(ls "$kernel_dir/lib/modules" || echo "Unknown")"
```

SuperSandro2000 commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -503,11 +516,85 @@ if [ "$action" = dry-build ]; then
+        kernel_dir="$(dirname "$(realpath "$generation_dir/kernel")")"
+        kernel_version="$(ls "$kernel_dir/lib/modules" || echo Unknown)"
+
+        configurationRevision="$("$generation_dir/sw/bin/nixos-version" --configurationRevision 2> /dev/null || :)"
```
```suggestion
        configurationRevision="$("$generation_dir/sw/bin/nixos-version" --configuration-revision 2> /dev/null || true)"
```

SuperSandro2000 commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -503,11 +516,85 @@ if [ "$action" = dry-build ]; then
+        # Old nixos-version output ignored unknown flags and just printed the version
+        # therefore the following workaround is done not to show the default output
+        nixos_version_default="$("$generation_dir/sw/bin/nixos-version")"
+        if [ "$configurationRevision" = "$nixos_version_default" ]; then
```
```suggestion
        if [ "$configurationRevision" == "$nixos_version_default" ]; then
```

SuperSandro2000 commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -503,11 +516,85 @@ if [ "$action" = dry-build ]; then
+            nixos_version="$(jq -aR <<< "$nixos_version")"
+            kernel_version="$(jq -aR <<< "$kernel_version")"
+            configurationRevision="$(jq -aR <<< "$configurationRevision")"
+            echo "{ \"generation\": $generation_number, \"date\": $build_date, \"nixosVersion\": $nixos_version, \"kernelVersion\": $kernel_version, \"configurationRevision\": $configurationRevision, \"current\": $current_generation_tag}"
```
Not sure if we could add a linebreak here somewhere.

Mic92 replied:
Maybe add a second `echo`.

miallo replied:
Yes - this line is completely incomprehensible... Maybe one could even put each entry in a separate line or is that too much?

miallo replied:
I now put it in separate lines joined with backticks - what do you think about that?

SuperSandro2000 commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -503,11 +516,85 @@ if [ "$action" = dry-build ]; then
+        done |
+        if [ -z "$json" ]; then
+            column --separator "," --table --table-columns "Generation,Build-date,NixOS version,Kernel,Configuration Revision" |
+                ${PAGER:less}
```
Do we always have less? I am not sure.

Mic92 replied:
I don't think so. Better to fallback to cat.

SuperSandro2000 reviewed (commented):
I am not sure if we use [[ or [ in the file. I think so far we are mixing.

Can you please run shellcheck and fix all reasonable complaints in the code you changed? I think there are some left.

And please remind me to merge that after the 22.05 branch off.

Note to myself:
- test this on my systems.
- send to nixos-dev channel to get more people to review it

Mic92 commented on pkgs/os-specific/linux/nixos-rebuild/default.nix:
```diff
@@ -19,7 +21,7 @@ substituteAll {
   nix_i686_linux = fallback.i686-linux;
   nix_aarch64_linux = fallback.aarch64-linux;
-  path = lib.makeBinPath [ coreutils gnused gnugrep ];
+  path = lib.makeBinPath [ coreutils gnused gnugrep jq util-linux ];
```
Interestingly jq is already in $PATH in my nixos-rebuild. Where does it come from when it's not in this list?

miallo replied:
Interesting... It previously was in this list, but it got removed with c75bc3abc760a07eb7afc0be26eb1813a9867a84 because it was no longer in use.

Mic92 commented on nixos/doc/manual/man-nixos-version.xml:
```diff
@@ -21,6 +21,10 @@
    </arg>
 
+   <arg>
+    <option>--configurationRevision</option>
```
as @SuperSandro2000 noted. I would changes this to `--configuration-revision` as this is more consistent with other nixos cli tools.

Mic92 commented:
This is a very long-running PR, but I think it's almost there. Just the few nitpicks above.

miallo commented:
> I am not sure if we use [[ or [ in the file. I think so far we are mixing.

Correct, e.g.: https://github.com/NixOS/nixpkgs/blob/2516caecaea15912cc7991e06e343f0b3045224a/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L163-L166
I am fine to switch it in any direction. Should I go for `[[` in this PR, because it is in there any ways?

> Can you please run shellcheck and fix all reasonable complaints in the code you changed? I think there are some left.

I think (now) there is nothing from this PR left. A few `SSHOPTS`, one `mkdir` from 2013 and something related to `nixos-rebuild edit`

> send to nixos-dev channel to get more people to review it
> Just the few nitpicks above.

Thanks a lot for the review already! I definitely can learn a lot more about shell scripting, so every comment is highly valued!

miallo commented:
@SuperSandro2000 and @Mic92 Should I squash the commits into the three separate topics aka `nixos-version`, `list-generations` and 'switch --generation` or do you want to review them as they are?

miallo commented:
Ping @Mic92 - I know you are very busy with lots of important stuff (thanks a lot for your involvement!). How can I help with this review process?

miallo commented:
Hi @Profpatsch!
Could you take a look at this PR? Thank you very much!

Mic92 commented:
I tested your changes but its seems that the normal output seems to not contain any output:

```
╭─[~/git/nixpkgs]─[main]
╰─ % sudo nixos-rebuild list-generations 
```

The json output works but it could have more newlines in it for readability:

```
[
{                 "generation": 561,                 "date": "2023-01-22T16:42:32Z",                 "nixosVersion": "23.05.20230122.e6c6668",                 "kernelVersion": "6.0.19",                 "configurationRevision": "",                 "current": true            },{                 "generation": 560,                 "date": "2023-01-22T16:00:57Z",                 "nixosVersion": "23.05.20230122.e6c6668",                 "kernelVersion": "6.0.19",                 "configurationRevision": "",                 "current": false            },{                 "generation": 559,                 "date": "2023-01-22T15:52:27Z",                 "nixosVersion": "23.05.20230122.e6c6668",                 "kernelVersion": "6.0.19",                 "configurationRevision": "",                 "current": false            },{                 "generation": 558,                 "date": "2023-01-22T10:33:12Z",                 "nixosVersion": "23.05.20230122.c625409",                 "kernelVersion": "6.0.19",                 "configurationRevision": "",                 "current": false            },{                 "generation": 557,                 "date": "2023-01-20T08:13:56Z",                 "nixosVersion": "23.05.20230118.ca65306",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 556,                 "date": "2023-01-18T16:27:39Z",                 "nixosVersion": "23.05.20230118.ca65306",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 555,                 "date": "2023-01-18T16:03:49Z",                 "nixosVersion": "23.05.20230118.ca65306",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 554,                 "date": "2023-01-18T15:17:52Z",                 "nixosVersion": "23.05.20230118.2b7d772",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 553,                 "date": "2023-01-18T12:30:17Z",                 "nixosVersion": "23.05.20230115.7105f0b",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 552,                 "date": "2023-01-16T15:03:07Z",                 "nixosVersion": "23.05.20230115.7105f0b",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 551,                 "date": "2023-01-15T22:34:56Z",                 "nixosVersion": "23.05.20230115.7105f0b",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 550,                 "date": "2023-01-15T17:00:30Z",                 "nixosVersion": "23.05.20230113.85bd130",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 549,                 "date": "2023-01-15T16:53:52Z",                 "nixosVersion": "23.05.20230113.85bd130",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 548,                 "date": "2023-01-15T15:04:32Z",                 "nixosVersion": "23.05.20230113.85bd130",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 547,                 "date": "2023-01-15T15:00:13Z",                 "nixosVersion": "23.05.20230113.85bd130",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 546,                 "date": "2023-01-15T14:59:38Z",                 "nixosVersion": "23.05.20230113.85bd130",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 545,                 "date": "2023-01-15T14:59:28Z",                 "nixosVersion": "23.05.20230113.85bd130",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 544,                 "date": "2023-01-15T14:57:59Z",                 "nixosVersion": "23.05.20230113.85bd130",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 543,                 "date": "2023-01-15T14:52:23Z",                 "nixosVersion": "23.05.20230113.85bd130",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 542,                 "date": "2023-01-13T14:42:47Z",                 "nixosVersion": "23.05.20230113.85bd130",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 541,                 "date": "2023-01-13T14:42:30Z",                 "nixosVersion": "23.05.20230113.85bd130",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 540,                 "date": "2023-01-13T12:59:13Z",                 "nixosVersion": "23.05.20230113.feab905",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 539,                 "date": "2023-01-13T09:40:10Z",                 "nixosVersion": "23.05.20230110.929142a",                 "kernelVersion": "6.0.17",                 "configurationRevision": "",                 "current": false            },{                 "generation": 538,                 "date": "2023-01-10T22:40:53Z",                 "nixosVersion": "23.05.20230104.4b68943",                 "kernelVersion": "6.0.16",                 "configurationRevision": "",                 "current": false            }]

```

Mic92 commented:
Let me know if you can not reproduce the empty output and I will provide more debug information.

SuperSandro2000 commented:
Kinda works for me. The table formatting is a bit off.

```
 ➜ nixos-rebuild list-generations
Generation  Build-date                               NixOS version          Kernel                   Configuration Revision
73                          Sun 2023-01-22 22:34:14                  23.05                  5.15.89                  0000000000000000000000000000000000000000000000000000000000000000                  (current)
72                          Sun 2023-01-22 20:25:42                  23.05                  5.15.89                  0000000000000000000000000000000000000000000000000000000000000000         
71                          Sun 2023-01-22 20:19:00                  23.05                  5.15.89                  0000000000000000000000000000000000000000000000000000000000000000         
70                          Sat 2023-01-21 23:40:38                  23.05                  5.15.88                  0000000000000000000000000000000000000000000000000000000000000000         
69                          Tue 2023-01-17 19:37:59                  23.05                  5.15.86                  0000000000000000000000000000000000000000000000000000000000000000         
68                          Tue 2023-01-17 02:07:50                  23.05                  5.15.86                  0000000000000000000000000000000000000000000000000000000000000000         
```

miallo commented:
I am sorry for not replying soon...
> Kinda works for me. The table formatting is a bit off.

@SuperSandro2000 Sorry - That is on me. I tried to make a string more readable in the code and accidentally added some space padding to a couple of entries... Consider it fixed...

> The json output works but it could have more newlines in it for readability:

@Mic92 Right... We should probably just pipe the output through `jq` once more - that would directly pretty-print it as well including "syntax" highlighting.

Also I just realized that the `column` command has a `--json` flag, so maybe we can get rid of quite some sort of duplicate code where we prepare the output. Though this would be at the cost of less machine readable output:
<details><summary>Example difference between `column --json` and "hand generated" json</summary>

`column --table` output:
```json
  {
    "generation": "725",
    "build-date": "Mon 2023-01-30 09:49:02",
    "nixos version": "23.05.20230130.1ea79b6",
    "kernel": "6.1.7",
    "configuration revision": "1d171c75fc5bd6d30833ea05789bd4eac1c68831  (current)"
  },
```

vs hand generated JSON

```json
  {
    "generation": 725,
    "date": "2023-01-30T08:49:02Z",
    "nixosVersion": "23.05.20230130.1ea79b6",
    "kernelVersion": "6.1.7",
    "configurationRevision": "1d171c75fc5bd6d30833ea05789bd4eac1c68831",
    "current": true
  },
```
</details>

Differences:
- keys containing spaces might be a bit annoying to work with in JSON
- generations are numbers, not strings
- Date uses machine- vs. human-readable format
- the field of `current` for the generation is separated into its own column (could also be included in the table), and now is a boolean

Obviously one could also do it the other way around and generate the table from the JSON:
<details><summary>Example difference between `jq '@tsv'` and "hand generated" table</summary>

 `| jq -r '.[] | [.generation, .date, .nixosVersion, .kernelVersion, .configurationRevision, .current] | @tsv' | column --separator $'\t' --table --table-columns "Generation,Build-date,NixOS version,Kernel,Configuration Revision"`
 
 ```
725         2023-01-30T08:49:02Z  23.05.20230130.1ea79b6  6.1.7   1d171c75fc5bd6d30833ea05789bd4eac1c68831  true
724         2023-01-26T15:08:56Z  23.05.20230124.5a740de  6.0.19  ac698fcffad2420d0ba555fdedaa1c9761350282  false
```
vs current
```
725         Mon 2023-01-30 09:49:02  23.05.20230130.1ea79b6  6.1.7   1d171c75fc5bd6d30833ea05789bd4eac1c68831  (current)
724         Thu 2023-01-26 16:08:56  23.05.20230124.5a740de  6.0.19  ac698fcffad2420d0ba555fdedaa1c9761350282
```
</details>
This version is most likely more versatile, because I am relatively sure that one can transform e.g. the boolean `current` to empty strings or `"(current)"` and probably one can also get the date into a more human readable format as well...

miallo commented:
I will fix the conflict in the man page, though it will take me a bit of time to get to know the new format and apply it properly... I hope I can get it done over the next week :)

SuperSandro2000 commented:
I just dropped the merge commits and did a rebase on master.

miallo commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -507,11 +520,97 @@ if [ "$action" = dry-build ]; then
+    }
+
+    find "$(dirname "$profile")" -regex "$profile-[0-9]+-link" |
+        sort -rg |
```
FIXME: I thought the `-g` would take care of 1 < 2 < 10, but it is just ordered by first digit.

miallo replied:
```
Generation  Build-date               NixOS version           Kernel  Configuration Revision
9           Tue 2023-06-20 15:03:54  23.11.20230618.09c81a9  6.3.8   dirty
8           Tue 2023-06-20 14:27:04  23.11.20230618.09c81a9  6.3.8   40004285a151ec03048c826bc82b5fa34cf73ff0
7           Tue 2023-06-20 14:14:45  23.11.20230618.09c81a9  6.1.34  dirty
6           Tue 2023-06-20 14:05:13  23.11.20230618.09c81a9  6.1.34  dirty
5           Tue 2023-06-20 13:16:19  23.11.20230618.09c81a9  6.3.8   dirty
4           Tue 2023-06-20 12:58:30  23.11.20230618.09c81a9  6.3.8   dirty
3           Tue 2023-06-20 11:53:36  23.05.1139.572d2693045  6.1.34
2           Tue 2023-06-20 10:03:19  23.05.1139.572d2693045  6.1.34
1           Mon 2023-06-19 16:05:32  23.05.1123.aaef163eac7  6.1.34
11          Tue 2023-06-20 18:52:13  23.11.20230618.09c81a9  6.3.8   0f020d3fbaae0e0972911a0a786891998c58ef03  (current)
10          Tue 2023-06-20 18:43:05  23.11.20230618.09c81a9  6.3.8   e6dc04fa457e207b19c62bb3b66dd2b1796a9417
```

miallo commented:
I just learned about `--specialisation` - that might also be something that I had to consider...

miallo commented:
I don't have a clue what people could be interested in from the specialisations in this overview - As a first step I just went with their names for each generation for now, but maybe that is not useful in real world scenarios... Who has an idea how to provide the most value?
```
Generation  Build-date               NixOS version           Kernel  Configuration Revision                    Specialisations
31          Thu 2023-07-13 23:18:21  23.11.20230713.f53f3fd  6.4.3   dirty                                     *                  (current)
30          Thu 2023-07-13 21:31:07  23.11.20230713.f6ecfeb  6.4.3   dirty                                     plasma
29          Thu 2023-07-13 14:16:13  23.11.20230713.f6ecfeb  6.4.3   a97084a553fe221d943911e0850c163c985534fe  *
```


On a different note: the `(current)` at the end of the line of the table is a bit out of view - what do you think of including it in he generation column like
```
Generation     Build-date               NixOS version           Kernel  Configuration Revision  Specialisations
31  (current)  Thu 2023-07-13 23:18:21  23.11.20230713.f53f3fd  6.4.3   dirty                   *
```
?

miallo commented:
Now that the less critical part of this PR (`list-generations` #250501) is merged, this is now has the single focus of allowing to switch to a specific generation. This is how I should have done it in the first place, which would have made reviews a lot easier. Sorry for that!

What would be the next possible steps for this one? I have never written a NixOS-test, but maybe I can take a look into them, if that would be something reasonable to push this one forward.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2687

mairs8 commented:
i thought that switching to specific generation is acheived by: `/nix/var/nix/profiles/system-N-link/bin/switch-to-configuration switch`.

Is this thread about adding a more convenient syntax, for example to switch to generation 5: `nixos-rebuild switch --rollback 5` ?

miallo commented:
> i thought that switching to specific generation is acheived by: `/nix/var/nix/profiles/system-N-link/bin/switch-to-configuration switch` where `N` is the generation to switch to.

Yes, I think so. This PR is equivalent to
```bash
sudo nix-env --switch-generation N -p /nix/var/nix/profiles/system
sudo /nix/var/nix/profiles/system/bin/switch-to-configuration switch
```
because that is also the way it was implemented for `--rollback`. I am not 100% sure about the implications - it might just be the use of a couple of programs from the later (potentially broken) generation if I am not mistaken. But maybe this is indeed just too complex for this and the reason it was used for `--rollback` was not to duplicate the logic of "How to find the previous generation?" in nixos-rebuild and nix-env 🤔 Someone with a better understanding of the details should probably take a close look at it...

> Is this thread about adding a more convenient syntax

Yes, I think having an easier syntax would be better than the current way (and also more discoverable, especially as soon as you add things like different profiles/specialisations).
> for example to switch to generation 5: `nixos-rebuild switch --rollback 5` ?

 I considered your suggested syntax, but I think it is a bit ambiguous if you want to rollback to a specific generation (as is the intention here) or a relative amount (so in your case e.g. from generation 100 to 95). That is why I thought a different syntax is more precise and it is also the one Eelco suggested: https://github.com/NixOS/nixpkgs/pull/105910#issuecomment-753948794

bzm3r reviewed (commented):
Everything here makes sense/looks good to me. I too would be interested in helping @miallo pushing this forward.

miallo commented:
> Everything here makes sense/looks good to me. I too would be interested in helping @miallo pushing this forward.

That would be nice!

I just pushed a rebase where the only change I made was the update of the `nixos-rebuild list-generations` output in nixos/doc/manual/administration/rollback.section.md since the format has changed between when I implemented it and wrote this doc and now...

miallo commented:
> the current manpage mentions switch --generation n which is not yet available so the situation is quite confusing!

Oh damn... -.- That one is totally on me. I extracted the `list-generations` feature into a separate MR and even though I was certain that I fixed the man-page in that MR at some point not to include that reference I must have messed it up...
I don't have a clue on how to push this MR, so in order to avoid confusion I'll remove that sentence in a different MR...

miallo commented:
I just added some tests. This was the first time I wrote any nixos tests and so this is just an adaptation of  [tests/nixos-rebuild-specialisations.nix](https://github.com/NixOS/nixpkgs/blob/739d72c5c2181fc202d1a64fd1d575d1bdaca722/nixos/tests/nixos-rebuild-specialisations.nix) - but I don't really know if this can be simplified any more...

@Mic92 @SuperSandro2000 What else do you think is necessary for this PR?

@SuperSandro2000 you mentioned a few things you wanted to do in https://github.com/NixOS/nixpkgs/pull/105910#pullrequestreview-973193522

SuperSandro2000 commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -631,7 +640,7 @@ if [ -z "$rollback" ]; then
         copyToTarget "$pathToConfig"
     fi
-else # [ -n "$rollback" ]
+elif [[ -n "$rollback" ]]; then
```
```suggestion
elif [ -n "$rollback" ]; then
```

SuperSandro2000 commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -430,7 +439,7 @@ SSHOPTS="$NIX_SSHOPTS -o ControlMaster=auto -o ControlPath=$tmpDir/ssh-%n -o Con
 # First build Nix, since NixOS may require a newer version than the
 # current one.
-if [[ -n "$rollback" || "$action" = dry-build ]]; then
+if [[ "$action" = dry-build ]]; then
```
```suggestion
if [ "$action" = dry-build ]; then
```

SuperSandro2000 reviewed (commented):
lets only use one style of test at least in the changed code

SuperSandro2000 commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -68,6 +69,14 @@ while [ "$#" -gt 0 ]; do
+      --generation)
+        if [ $# -eq 0 ]; then
+            log "Must provide a generation number"
+            log "Usage: \`$(basename "$0") ${action:-switch} --generation <NUMBER>'"
```
Why do I need to give an action when I just want to switch to a generation?

miallo replied:
We don't know if you want to switch, boot or test said generation, so you have to tell us what you want to do with it...

SuperSandro2000 replied:
but if I want to switch to an older generation, why does it start to build the current config file from disk which might be newer than the current generation?

miallo replied:
That is interesting - I don't think this happened to me before :thinking: Where did you see this behavior? If it is in the tests: the first three tests don't actually do the `--generation`, but they create the generations, so there it is expected that they build the config files...

I know it isn't a good answer, but: I don't really know what I am doing (at least not 3 years after I first wrote this patch). I just adapted the code for the `--rollback` since I thought they would be more or less identical... I have forgotten the few things I knew about the nix-env stuff and the rest on how this is activated :see_no_evil: If we just wanted to activate the generation and don't care about the action I guess we could also just run `/nix/var/nix/profiles/system-$generation-link/activate`...

I tried to add a test for the `boot --generation`, but something is strange (maybe the QEMU runs some kind of tempfs - at least after the `machine.reboot()` and it doesn't have nix-channels or actually the old generations any more on the disk) and so I could only check that the current generation was still active...

miallo replied:
I could not reproduce the building of the current configuration.nix in the tests, as this passes, which I think is what you mean:
```diff
diff --git a/nixos/tests/nixos-rebuild-switch.nix b/nixos/tests/nixos-rebuild-switch.nix
index 8fbd1f2eb587..c726dcca7b83 100644
--- a/nixos/tests/nixos-rebuild-switch.nix
+++ b/nixos/tests/nixos-rebuild-switch.nix
@@ -89,7 +89,9 @@ import ./make-test-python.nix ({ pkgs, ... }: {
           ${testForActiveGeneration "3"}

       with subtest("must roll back with `switch --generation`"):
-          machine.succeed("nixos-rebuild switch --generation 1")
+          ${createConfig "4"}
+          out = machine.succeed("nixos-rebuild switch --generation 1")
+          assert "building the system configuration..." not in out
           ${testForBootGeneration "1"}
           ${testForActiveGeneration "1"}
```
Did I misunderstand you or do you have a reproduction by any chance?

xaverdh replied:
well the script (re)executes itself https://github.com/NixOS/nixpkgs/blob/1860f2ab371659d417f30e9f32e242c6ce5928a5/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L405C68-L405C75, so maybe that is why it starts building?

SuperSandro2000 replied:
> If we just wanted to activate the generation and don't care about the action I guess we could also just run `/nix/var/nix/profiles/system-$generation-link/activate`...

Yes, this is basically what I wanted to do.

> Did I misunderstand you or do you have a reproduction by any chance?

I think this is what I did, yes.

> well the script (re)executes itself

I am usually passing --fast to it.

miallo replied:
> > If we just wanted to activate the generation and don't care about the action I guess we could also just run `/nix/var/nix/profiles/system-$generation-link/activate`...
> 
> Yes, this is basically what I wanted to do.

I just implemented it with "feature parity" to the `--rollback`. I personally mostly use `switch` as an action, but maybe e.g. someone could want to set a "known good generation" as a default after a reboot when testing a server update and on error they can just do a reboot and everything should be back to normal. Could something like this be a use case (and if not, just out of curiosity: why would it be one for rollback)?

> > Did I misunderstand you or do you have a reproduction by any chance?
> 
> I think this is what I did, yes.

Very interesting... This should not happen (https://github.com/NixOS/nixpkgs/pull/105910/files#diff-fc23c722779993fcffc932965cd143b265aa196d577ca10173b63e5bdcbc0001R608-R609) and also the test that I wrote above passes... 🤔 I really don't have a clue how this could have happened... Anyone has any idea (or even better can add a failing test case)?

theduke commented:
Any hopes for pushing this forward?
Would be great to have this available.

miallo commented:
I rebased it and fixed the merge conflicts. Also added a tiny improvement to the test so that it is checked if a new genertion is built, or if it is just a simple switch

nyabinary commented:
> I rebased it and fixed the merge conflicts. Also added a tiny improvement to the test so that it is checked if a new genertion is built, or if it is just a simple switch

another merge conflict :P

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/where-did-you-get-stuck-in-the-nix-ecosystem-tell-me-your-story/31415/72

Creative-Difficulty commented:
Are there any updates on this?

Jayman2000 reviewed (changes requested):
The most important parts of this pull request are the changes to `nixos/doc/manual/administration/rollback.section.md`, `nixos/doc/manual/release-notes/rl-2411.section.md` and `pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh`. Those changes are good. There are some minor improvements that I would make to those files, but for the most part they’re good.

That being said, `nixos/tests/nixos-rebuild-switch.nix` needs work. I tried following [the instructions in the NixOS manual for running NixOS tests](https://nixos.org/manual/nixos/unstable/#sec-running-nixos-tests), but I wasn’t able to run the test by following those instructions.

I created [a branch named `nixos-rebuild/switch-generation-proposals`](https://codeberg.org/JasonYundt/nixpkgs-backup/src/branch/nixos-rebuild/switch-generation-proposals). Here’s what that branch contains:

- A rebased version of the commit from this pull request. I rebased it on a newer revision of the `master` branch (3a160317b15c707539d4211a51c83e01b3bbf370).
- A bunch of [fixup commits](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---fixupamendrewordcommit) that bring `nixos/tests/nixos-rebuild-switch.nix` into working order.
- A few fixup commits that make minor improvements to the other parts of this pull request.

Each fixup commit contains a detailed explanation for why I think that the change from the fixup commit should be made. Here’s what you can do to review my fixup commits and incorporate some or all of them into this pull request:

<details>
<summary>Click here to show or hide instructions.</summary>

1. Change directory into a clone of the Nixpkgs repository by running this command:

    ```bash
    cd <path to nixpkgs repo>
    ```

1. Make sure that your local copy of the Nixpkgs’s `master` branch is sufficiently up-to-date by running this command:

    ```bash
    git switch master && git pull
    ```

1. Create a local copy of my `nixos-rebuild/switch-generation-proposals` branch by running this command:

    ```bash
    git remote add codeberg-JasonYundt https://codeberg.org/JasonYundt/nixpkgs-backup.git && git fetch codeberg-JasonYundt && git switch --track codeberg-JasonYundt/nixos-rebuild/switch-generation-proposals
    ```

1. Read through the fixup commits and the code that they change. I recommend using `git log --reverse --patch master..` to read through the commits, but it doesn’t really matter how you do it.

1. If there are any fixup commits that you don’t like, then use `git rebase --interactive` to drop those fixup commits.

1. If there are any fixup commits that you do like, then incorporate them into this pull request by following these instructions:

    1. Figure out the hash for the first commit in the `nixos-rebuild/switch-generation-proposals` branch by running this command:

        ```bash
        git cherry -v master HEAD | head --lines=1
        ```

    1. Squash the commits in your local copy of the `nixos-rebuild/switch-generation-proposals` branch by running this command:

        ```bash
        git rebase --autosquash <hash>^
        ```

    1. Switch to your local `nixos-rebuild/switch-generation` branch by running this command:

        ```bash
        git switch nixos-rebuild/switch-generation
        ```

    1. Make your local `nixos-rebuild/switch-generation` branch the same as your local `nixos-rebuild/switch-generation-proposals` branch by running this command:

        ```bash
        git reset --hard nixos-rebuild/switch-generation-proposals
        ```

    1. Push a new version of this pull request by running this command:

        ```bash
        git push --force
        ```

</details>

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 105910`
Commit: `79c579c37c119892647875fb53bb69f98e7d39b4`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixos-rebuild</li>
  </ul>
</details>

eclairevoyant commented:
I'm wondering if it makes more sense to invest time in improving the -ng script since that's really where other improvements are being made.

thiagokokada reviewed (changes requested):
Let's not merge this unless this is reworked to use nixos-rebuild-ng instead.
